### ISSUE TITLE
model.predict is much slower on TF 2.1+

### ISSUE THREAD

**Issue by lihanchen** [tensorflow/tensorflow#40261]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): WIndows 10 and Ubuntu 18.04
- TensorFlow installed from (source or binary): Binary with pip3
- TensorFlow version (use command below): 2.1+ vs. 2.0
- Python version: 3.7
- CUDA/cuDNN version: Used with CPU
- CPU model: Intel i7 5930

**Describe the current behavior**
Starting from tensorflow-cpu 2.1, my program spends multiple fold of time on model.predict() compared to tensorflow 2.0. TF 2.2 get about the same result as 2.1.
My original program is fairly complicate. I wrote a simpliest example code below.
With TF 2.0, it takes 0.13 seconds to run.
With TF 2.2, it takes about 3 seconds.

**Describe the expected behavior**
It should have similar execution time with TF 2.1+

**Standalone code to reproduce the issue**
```
from tensorflow.keras import Input, Model
import time
import numpy as np

x = Input(shape=(1, 1))
model = Model(inputs=x, outputs=x)

t = time.time()
i = 0
while i<100:
    model.predict(np.zeros((1, 1, 1)))
    i += 1
print(time.time() - t)
```

**Comment by Saduf2019** [comment]
@
I ran the code shared by you on [tensorflow 2.0](https://colab.sandbox.google.com/gist/Saduf2019/490e7394f67121bc4997845b4491c610/untitled216.ipynb), [tf 2.1](https://colab.sandbox.google.com/gist/Saduf2019/54e2392e40ed2afd588ede7cd298a7ca/untitled208.ipynb) and [tf 2.2 ](https://colab.sandbox.google.com/gist/Saduf2019/47f16b6ecdd34f22e019f8ab50ad61ac/untitled216.ipynb) and do not see much of a difference.

**Comment by lihanchen** [comment]
Hi, I just tried with another machine with i7-6700 and Linux Mint 19. I can repro it.
Did you run it on CPU only? Also, since tf 2.1 has a difference package name (tensorflow-cpu vs tensorflow), you need to uninstall the other one or use virtualenv.

```
REMOTE lhc@LHC-DEV-Desktop  ~/venv  cd tf2.0/bin
REMOTE lhc@LHC-DEV-Desktop  ~/venv/tf2.0/bin  source activate
(tf2.0) REMOTE lhc@LHC-DEV-Desktop  ~/venv/tf2.0/bin  cd -
~/venv
(tf2.0) REMOTE lhc@LHC-DEV-Desktop  ~/venv  python3 test.py
2020-06-08 14:59:32.578253: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2020-06-08 14:59:32.599598: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 3399905000 Hz
2020-06-08 14:59:32.600206: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x3ffd1f0 executing computations on platform Host. Devices:
2020-06-08 14:59:32.600234: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): Host, Default Version
0.09805583953857422
(tf2.0) REMOTE lhc@LHC-DEV-Desktop  ~/venv  python3 test.py
2020-06-08 14:59:36.514623: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2020-06-08 14:59:36.535721: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 3399905000 Hz
2020-06-08 14:59:36.536023: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x588ddf0 executing computations on platform Host. Devices:
2020-06-08 14:59:36.536045: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): Host, Default Version
0.09672117233276367
(tf2.0) REMOTE lhc@LHC-DEV-Desktop  ~/venv  cd tflatest/bin
(tf2.0) REMOTE lhc@LHC-DEV-Desktop  ~/venv/tflatest/bin  source activate
(tflatest) REMOTE lhc@LHC-DEV-Desktop  ~/venv/tflatest/bin  cd -
~/venv
(tflatest) REMOTE lhc@LHC-DEV-Desktop  ~/venv  python3 test.py
2020-06-08 14:59:52.216610: I tensorflow/core/platform/cpu_feature_guard.cc:143] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2020-06-08 14:59:52.239617: I tensorflow/core/platform/profile_utils/cpu_utils.cc:102] CPU Frequency: 3399905000 Hz
2020-06-08 14:59:52.239968: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x47f95c0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-06-08 14:59:52.239992: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
1.5681874752044678
(tflatest) REMOTE lhc@LHC-DEV-Desktop  ~/venv  python3 test.py
2020-06-08 14:59:57.516289: I tensorflow/core/platform/cpu_feature_guard.cc:143] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2020-06-08 14:59:57.539687: I tensorflow/core/platform/profile_utils/cpu_utils.cc:102] CPU Frequency: 3399905000 Hz
2020-06-08 14:59:57.540030: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x334c360 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-06-08 14:59:57.540052: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
1.567026138305664

```
I also zipped the entire directory of testing code and 2 venvs.
https://drive.google.com/file/d/1G4GqDAi0z2jscEc7WSVcMxHr-dyr5QZc/view?usp=sharing

**Comment by ymodak** [comment]
I can reproduce this with mac OS Catalina version 10.15.
TF 2.0 took 0.116 seconds.
TF 2.2 ran for 2.09 seconds.

@lihanchen The colab links given by @Saduf2019 use cpu accelerator also the cpu and gpu packages are combined in `tensorflow` unless we specifically flag it as `tensorflow-cpu` or `tensorflow-gpu` in TF 2.1 and above.

**Comment by lihanchen** [comment]
Any updates?

**Comment by tomerk** [comment]
Hi @lihanchen, judging by your filed bug and your example code, I'm assuming you're running `model.predict` inside of a loop?

Model.predict is a top-level API designed for batch-predicting outside of any loops, with the fully featureset of the Keras apis. This means it manages things like converting your input to a tf.dataset and batching it, putting your computation into a tf.function, handling keras callbacks, etc.

If you're looking for a quick low-overhead model call to put inside of a loop / inside your own tf.function, we suggest directly calling the model on your data instead (w/ `training=True` to put the model in inference mode)

For example, running the following in colab with the tf.nightlies:
```
x = Input(shape=(1, 1))
model = Model(inputs=x, outputs=x)

t = time.time()
i = 0
while i<100:
    model.predict(np.zeros((1, 1, 1)))
    i += 1
print(time.time() - t)
```
prints `3.521230459213257`

```
x = Input(shape=(1, 1))
model = Model(inputs=x, outputs=x)

t = time.time()
i = 0
while i<100:
    model(np.zeros((1, 1, 1)), training=False)
    i += 1
print(time.time() - t)
```
prints `0.01329183578491211`

As you can see there's a 300x difference in the constant overheads.

-------------------------
All that being said:

1. If I'm mistaken about how you're actually using predict (and you're using it outside of any loops), please let us know!

If you can point us to an example (or input data) where a single predict call that's actually processing multiple batches of data (& possibly using various keras `predict` functionality) is much slower than in 2.0 that would be super helpful.

2. 35 milliseconds / `predict` seems like a shockingly high overhead for `predict`, I'll look into this. It's unlikely to make it into 2.3 unless we cherrypick, but I'll see what can be done.

**Comment by lihanchen** [comment]
Hi. Thank you for taking time. After changing from model.predict() to model().
The execution take much shorter time with above model.

My real model is a GRU followed by a few dense layers.
Here is the example code:
```
from tensorflow.keras import Input, Model
from tensorflow.keras.layers import GRU, Dense, ReLU, Lambda
import time
import numpy as np

x = Input(shape=(15, 60))

rnn = GRU(50, dropout=0.5, recurrent_dropout=0.5)(x)

y = Dense(1)(rnn)

model = Model(inputs=x, outputs=y)

t = time.time()
i = 0
while i<100:
    model.predict(np.random.rand(1, 15, 60))
    i += 1
print(time.time() - t)
```

I found some interesting result:
1. with TF 2.2, model.predict() call, it takes about 3.5 seconds with or without both GRU dropouts
2. with TF 2.2, model() call, it takes about 1.2 seconds without both GRU dropouts
3. with TF 2.2, model() call, it takes about 5 seconds with both GRU dropouts
4. with TF 2.0, model.predict() call, it takes less than 0.6 seconds with or without both GRU dropouts

Basically TF 2.0 is still much faster even I use model() cal to predict. 3 seems like a bug to me as dropouts make model() call slower than model.predict(). Is there a way I can achieve close to TF 2.0 performance with my model with TF 2.2?

**Comment by tomerk** [comment]
What exactly does your workload look like? Is it a batch predict situation? Are you serving a lot of very small predictions interactively?

If you have a batch-predicting situation where you have a lot of data you're trying to predict on, I suggest loading it with `tf.data` datasets and passing it directly to `.predict` instead of calling `.predict` in a loop. This is the setting `model.predict` is optimized for performance-wise

`model.predict` automatically wraps your model in a `tf.function`, which will generally improve performance for all but the smallest models.
If you want the advantage of a `tf.function` without the other overheads of `predict` for an environment where you're interactively predicting, then I would suggest defining a tf.function that calls your model, e.g.

```
@tf.function
def serve(x):
  return model(x, training=False)
```

**Comment by lihanchen** [comment]
It is a stock prediction software. Every hour, I feed in the corrent features into RNN and hopefully it will generate some output as a prediction of stock trend in next time period.
I want the computation time to be minimum so the software can submit order based on current knowledge as soon as possible.

**Comment by tomerk** [comment]
I see.

After doing some tracing of `model.predict` in the no-op model, almost all the time is spent in a constant overhead when converting the single numpy value to a tf.data dataset. If you were to convert your entire dataset to a tf.data.dataset once at the start (outside of the loop), you would only have to pay that 30ms cost once.

So, if need to minimize latency on a very very small amount of data, I'd suggest using the `tf.function`'d serve method I mentioned above. This will minimize any constant-time overheads while still tf.functioning & optimizing your model.

If you expect your operations to take longer & your data is bigger, I'd suggest calling `.predict` once on your whole dataset outside of any loops. `predict` will be able to parallelize your input pipeline for you which may produce better throughput than just calling `serve` in a loop you've written yourself.

**Comment by lihanchen** [comment]
Thanks for investigating.
30ms to covert an 1x1x1 numpy sounds a long time but with no-op model I can achieve short time with model() call.
The question is why dropouts take much long time and why TF2.0 is much faster.

**Comment by tomerk** [comment]
Creating datasets comes with a constant overhead regardless of the data size. We may be able to improve it in the future for the case of small numpy data, but it's unlikely to improve real use cases so it's not a high priority.

Do your dropouts take a long time/is tf 2.0 faster when you `tf.function` the forward pass as mentioned in the above `serve` method?

**Comment by lihanchen** [comment]
I didn't use tf.data before. I will learn it and try.

**Comment by lihanchen** [comment]
When I change it to create a dataset in the loop, it is much faster
```
while i<100:
    train_dataset = tf.data.Dataset.from_tensor_slices(np.random.rand(1, 15, 60)).batch(1)
    model.predict(train_dataset)
    i += 1
```
It take about 150% - 200% time of TF 2.0. I think there are still rooms to optimize as I create the data set in the loop, but I am satisfied with current solution.

**Comment by tangcxx** [comment]
I have met the same problem. I solved this by downgrading keras 2.4.3 to 2.3.1 while I keep tf as 2.2.0.

**Comment by ectg** [comment]
I see a huge difference between TF 1.13 and TF 2.2 — running predict on a single image now takes 0.05 seconds, it was 0.01 seconds on TF 1.13 (25 fps vs 100 fps on video data). 

Using model(x) instead of model.predict(x, training=False) further degrades the performance.

Does anyone have a solution?

**Comment by tomerk** [comment]
@ectg Did you try tf.function-ing your `model.call` method before calling your model?
As such:

```
model.call = tf.function(model.call, experimental_relax_shapes)
model(..., training=False)
```

Also, what are the use cases where you all are finding that building a `tf.dataset` at the start and batch predicting is impractical?
Knowing this would help our prioritization.

**Comment by ectg** [comment]
@tomerk I just tried out your suggestion. The frame rate has gone up from 48 fps to 253 fps. Thanks! 

As to building a tf.dataset at the start -- the network is intended for real time object detection on a video stream, so the dataset is not available beforehand.

**Comment by haha1227** [comment]
@tomerk This is the exact solution. Thanks. I guess this is a bug you should change

**Comment by piyohogeo** [comment]
@tomerk Thanks! I replaced many "predict" in my projects. I think this is a critical performance issue.

**Comment by owni1337** [comment]
Hi,

i got a Problem with the returnValue from model(xxx, training=False).
I have a working Program with predict.

`valueRed    = float(col.r)
valueGreen  = float(col.g)
valueBlue   = float(col.b)
data = np.array([[valueRed/255, valueGreen/255, valueBlue/255]])  
result = model.predict(data, batch_size=32)
index = result.argmax(1)
return (labelList[int(index)])`

The Result is a numpy:
[[1.8138054e-03 5.4888315e-03 6.9668889e-04 2.9702934e-03 8.9826369e-01
  9.0765595e-02 1.0664122e-06]]


if i change the predict to model(xxx, training=False) it will be much faster but the Result isn't a numpy anymore and i can't use argmax on result.
The result looks like:
tf.Tensor(
[[1.8138054e-03 5.4888315e-03 6.9668889e-04 2.9702934e-03 8.9826369e-01
  9.0765595e-02 1.0664122e-06]], shape=(1, 7), dtype=float32)

How can i easy handle it to get my code working?

thanks a lot

**Comment by cccat6** [comment]
@tomerk Hi tomerk,

I know it might be an old issue, but it still exists today. I am using `tensorflow-gpu 2.6.0` with `keras 2.6.0`.
These two lines of code have the same output but the `predict` function spends much much more time than `model`.
```
self.model.predict(state) # slow
self.model(state, training=False) # fast
```
The output type of `predict` is a Python Array, while `model` is a TensorFlow Array. I don't think there is any actual difference. Please forgive me I did not view the code in these parts but just throw the question.
Just to inform that it should be an issue since the time difference is huge but the results are the same.

There is also a strange part but might be my problem. I feel like the `predict` function did not use my GPU since the usage of GPU is 0% when it calling. Training does use GPU.

Thanks a lot.

**Comment by tomerk** [comment]
@owni1337  sorry if this is too late, but you can grab the numpy value of an eager tensor `tensor.numpy()`

@cccat6 As mentioned above, `model.predict` Model.predict is a top-level API designed for batch-predicting outside of any other loops. Because it is designed for batch prediction and comes with other functionality, it comes with an inherently higher overhead. Much of this overhead is python or cpu-side overhead and is unrelated to the actual model computation.

I'm going to go ahead and close this issue for now, as we have added to the model.predict documentation & I've opened the above PR to update the docstring further.

**Comment by pty0220** [comment]
Hello, everyone!

I would like to get prediction results from my trained model as fast as possible 

Thus, I tried several combinations which suggested previously 
and the speed result is look like this (left is fastest)

`import tf1  result = model(input)` > `import tf2  result = model(input)` >= `import tf1  result = model.predict(input)`

Especially, `import tf1  result = model(input)` is super faster than the other methods
However,  in tf1, converting "result" tensor to numpy array using (sess.run() or eval and other way like No.1) is very time consuming 

Thus, I would be really appreciated it if you guys give some opinions or tips to solve this problem. 

This is my codes 
In the code function name "AI_simulation" is the callback function 

In the code result 
No.1 -> 4.1 FPS ~3.6 FPS getting slow (I think there is some memory leak)
No.2 -> 1.6 FPS (sess.run() is very time consuming )
No.3 -> 3.9 FPS
No.4 -> 3.8 FPS


No.1 
```

import tensorflow.compat.v1 as tf
tf.disable_eager_execution()
tf.disable_v2_behavior()

sess = tf.Session()
with sess.as_default():

    G_A2B = model_from_json(loded_model_json, custom_objects={'ReflectionPadding3D':ReflectionPadding3D, 'InstanceNormalization':InstanceNormalization})
    G_A2B.load_weights(model_path+'G_A2B_model_epoch_'+str(model_num)+'.hdf5')
    place_holder = tf.compat.v1.placeholder(tf.float32, shape=(1,170,110,110,1))
    G_A2B = G_A2B(place_holder, training=False)

def AI_simulation(AI_input):

    result = sess.run(G_A2B, feed_dict={place_holder: AI_input[np.newaxis, :, :, :, np.newaxis]})
    result = np.squeeze(result)

    return result
```
 

No.2 
```

import tensorflow.compat.v1 as tf
tf.disable_eager_execution()
tf.disable_v2_behavior()

sess = tf.Session()
with sess.as_default():

    G_A2B = model_from_json(loded_model_json, custom_objects={'ReflectionPadding3D':ReflectionPadding3D, 'InstanceNormalization':InstanceNormalization})
    G_A2B.load_weights(model_path+'G_A2B_model_epoch_'+str(model_num)+'.hdf5')


def AI_simulation(AI_input):

    result = G_A2B(AI_input[np.newaxis, :, :, :, np.newaxis])
    result = sess.run(result )
    result = np.squeeze(result)

    return result
```



No.3

```
import tensorflow.compat.v1 as tf
tf.disable_eager_execution()
tf.disable_v2_behavior()

G_A2B = model_from_json(loded_model_json, custom_objects={'ReflectionPadding3D':ReflectionPadding3D, 'InstanceNormalization':InstanceNormalization})
G_A2B.load_weights(model_path+'G_A2B_model_epoch_'+str(model_num)+'.hdf5')
G_A2B.call = tf.function(G_A2B.call, experimental_relax_shapes = True)

def AI_simulation(AI_input):

    result = G_A2B.predict(AI_input[np.newaxis, :, :, :, np.newaxis])
    result = np.squeeze(result)
    return result

```



No.4 

```
import tensorflow as tf  #tf2

G_A2B = model_from_json(loded_model_json, custom_objects={'ReflectionPadding3D':ReflectionPadding3D, 'InstanceNormalization':InstanceNormalization})
G_A2B.load_weights(model_path+'G_A2B_model_epoch_'+str(model_num)+'.hdf5')
G_A2B.call = tf.function(G_A2B.call, experimental_relax_shapes = True)

def AI_simulation(AI_input):
    result = G_A2B(AI_input[np.newaxis, :, :, :, np.newaxis])
    result = np.squeeze(result.numpy())
    return result

```
